notecard/main.go

```diff
@@ -1,107 +1,93 @@
 package main
 
 import (
-	"encoding/json"
 	"fmt"
+	"io"
+	"log"
 	"net/http"
 	"os"
 
 	"github.com/blues/note-go/notecard"
 )
 
-type Transport string
-
 const (
-	Serial Transport = "serial"
-	I2C    Transport = "i2c"
+	Serial = "serial"
+	I2C    = "i2c"
 )
 
-var card *notecard.Context
+type server struct {
+	card *notecard.Context
+}
 
-func serveNotecard(w http.ResponseWriter, req *http.Request) {
+func (s *server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if s.card == nil {
+		log.Fatal("notecard not initialized")
+	}
 
 	if req.Method != http.MethodPost {
 		w.WriteHeader(http.StatusMethodNotAllowed)
 		fmt.Fprintf(w, "Method not allowed")
 		return
 	}
 
-	var data map[string]interface{}
-	if err := json.NewDecoder(req.Body).Decode(&data); err != nil {
-		http.Error(w, fmt.Sprintf("Error decoding JSON: %v", err), http.StatusBadRequest)
-		return
-	}
-
-	for key, value := range data {
-		fmt.Printf("%s: %v\n", key, value)
-	}
-
-	jsonString, err := json.Marshal(data)
+	body, err := io.ReadAll(req.Body)
 	if err != nil {
-		fmt.Println("Error encoding JSON:", err)
-		w.WriteHeader(http.StatusInternalServerError)
-		panic(err)
+		log.Printf("while reading request body: %v", err)
+		http.Error(w, "error reading request body", http.StatusInternalServerError)
+		return
 	}
+	req.Body.Close()
 
 	// Print the resulting JSON string
-	fmt.Println(string(jsonString))
-
-	note_rsp, err := card.Transaction(data)
-	if err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
-		panic(err)
-	}
+	fmt.Println(string(body))
 
-	note_rsp_json, err := json.Marshal(note_rsp)
+	note_rsp, err := s.card.TransactionJSON(body)
 	if err != nil {
-		fmt.Println("Error encoding JSON:", err)
 		w.WriteHeader(http.StatusInternalServerError)
-		panic(err)
+		log.Printf("while doing a card transaction: %v", err)
+		return
 	}
 
 	// Set the Content-Type header to application/json
 	w.Header().Set("Content-Type", "application/json")
-	w.Write(note_rsp_json)
-
-	w.WriteHeader(http.StatusOK)
+	w.Write(note_rsp)
```

Comment:
this can return an error (network error for ex). it won't of course change what we serve in that case (because it's too late), but we could/should still log about it if it happens.